### PR TITLE
chore(tests): migrate to vitest

### PR DESCRIPTION
**Pull request by arjunvegda**
## Summary
- Migrate to vitest
- Following [pmndrs/jotai#1839](https://github.com/pmndrs/jotai/pull/1839)

## Check List

- [x] `yarn run prettier` for formatting code and docs
- [x] Wait for #1678 to be merged and update tests if needed
- [x] Ensure that the old TypeScript tests work as expected

### CODE REVIEW

File: vitest.config.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    name: 'zustand',
+    // Keeping globals to true triggers React Testing Library's auto cleanup
+    // https://vitest.dev/guide/migration.html
+    globals: true,
+    environment: 'jsdom',
+    dir: 'tests',
+    reporters: 'basic',
+    coverage: {
+      reporter: ['text', 'json', 'html', 'text-summary'],
+      reportsDirectory: './coverage/',
```
Comment (dammy001) [inline vitest.config.ts:20]:
```suggestion
      reportsDirectory: './coverage/',
      provider: 'c8',
```

**Comment by dai-shi** [comment]
@arjunvegda #1678 is merged.

File: vitest.config.ts (hunk 1 above)
Comment (arjunvegda) [reply vitest.config.ts:20]: `c8` is a default provider so we should be good here.

File: .github/workflows/test-multiple-builds.yml [hunk 2]
```diff
@@ -69,7 +69,6 @@ jobs:
         run: |
           sed -i~ "s/<rootDir>\/src\(.*\)\.ts/<rootDir>\/dist\/umd\1.development.js/" package.json
           sed -i~ 's/"test:ci":.*,$/"test:ci": "jest",/' package.json
-          sed -i~ 's/= await import(/= require(/' tests/devtools.test.tsx
```
Comment (dai-shi) [inline .github/workflows/test-multiple-builds.yml:72]: should remove L80 too?

**Review by dai-shi — COMMENTED** [review]
looks good except for one.

File: .github/workflows/test-multiple-builds.yml (hunk 2 above)
Comment (arjunvegda) [reply .github/workflows/test-multiple-builds.yml:72]: Good call! [Done](https://github.com/pmndrs/zustand/pull/1753/commits/3befb2bdbbae1687faf492f3dad873594a5e6920)

Comment (dai-shi) [reply .github/workflows/test-multiple-builds.yml:72]: I didn't realize that one.
I meant this: `sed -i~ 's/= await import(/= require(/' tests/devtools.test.tsx`.

Comment (arjunvegda) [reply .github/workflows/test-multiple-builds.yml:72]: Oh! I'm surprised that it did not cause the tests to fail this time 🤔 Fixed it in [this commit](https://github.com/pmndrs/zustand/pull/1753/commits/f5560adbdcce5e029dc87f34919d8038ebf6e435)

**Review by dai-shi — APPROVED** [review]
LGTM

File: package.json [hunk 3]
```diff
@@ -132,44 +130,9 @@
...
-      "text-summary"
-    ],
-    "collectCoverageFrom": [
-      "src/**/*.{js,ts,tsx}",
-      "tests/**/*.{js,ts,tsx}"
-    ]
-  },
   "dependencies": {
+    "@vitest/ui": "^0.30.1",
+    "eslint-plugin-vitest": "^0.1.4",
```
Comment (dai-shi) [inline package.json:135]: I didn't notice when reviewing, but these shouldn't be in dependencies.

Comment (dai-shi) [reply package.json:135]: https://github.com/pmndrs/zustand/pull/1789/commits/a0090b79008d044df78e07b35bfd5574d1382543

**Pull request merged**